sycl/source/detail/device_global_map_entry.cpp

```diff
@@ -53,6 +53,79 @@ OwnedUrEvent DeviceGlobalUSMMem::getInitEvent(adapter_impl &Adapter) {
   }
 }
 
+bool DeviceGlobalMapEntry::isAvailableInContext(const context_impl *CtxImpl) {
+  std::lock_guard<std::mutex> Lock{MDeviceToUSMPtrMapMutex};
+  for (const auto &It : MDeviceToUSMPtrMap)
+    if (It.first.second == CtxImpl)
+      return true;
+  return false;
+}
+
+bool DeviceGlobalMapEntry::isProfileCounter() {
+  const std::string CounterPrefix = "__profc_";
+  return MUniqueId.substr(0, CounterPrefix.size()) == CounterPrefix;
+}
+
+#ifdef _MSC_VER
+extern "C" void
+__sycl_increment_profile_counters(std::uint64_t FnHash, std::size_t NumCounters,
+                                  const std::uint64_t *Increments);
+extern "C" void
+__sycl_increment_profile_counters_default(std::uint64_t FnHash,
+                                          std::size_t NumCounters,
+                                          const std::uint64_t *Increments) {
+  (void)FnHash;
+  (void)NumCounters;
+  (void)Increments;
+}
+#pragma comment(                                                               \
+    linker,                                                                    \
+    "/alternatename:__sycl_increment_profile_counters=__sycl_increment_profile_counters_default")
+#else
+extern "C" void __attribute__((weak))
+__sycl_increment_profile_counters(std::uint64_t FnHash, std::size_t NumCounters,
+                                  const std::uint64_t *Increments);
+#endif
+
+void DeviceGlobalMapEntry::cleanupProfileCounter(context_impl *CtxImpl) {
+  std::lock_guard<std::mutex> Lock{MDeviceToUSMPtrMapMutex};
+  const std::size_t NumCounters = MDeviceGlobalTSize / sizeof(std::uint64_t);
+  const std::uint64_t FnHash = [&] {
+    const auto PrefixSize = std::string{"__profc_"}.size();
+    constexpr int DecimalBase = 10;
+    return std::strtoull(MUniqueId.substr(PrefixSize).c_str(), nullptr,
+                         DecimalBase);
+  }();
+  for (const device_impl &Device : CtxImpl->getDevices()) {
+    auto USMPtrIt = MDeviceToUSMPtrMap.find({&Device, CtxImpl});
+    if (USMPtrIt != MDeviceToUSMPtrMap.end()) {
+      DeviceGlobalUSMMem &USMMem = USMPtrIt->second;
+
+      // Get the increments from the USM pointer.
+      std::vector<std::uint64_t> Increments(NumCounters);
+      const std::uint64_t *Counters = static_cast<std::uint64_t *>(USMMem.MPtr);
+      for (std::size_t I = 0; I < NumCounters; ++I)
+        Increments[I] = Counters[I];
```

Comment:
CUDA and L0 adapters support `urUSMContextMemcpyExp`, which could allow this to also use device allocations. The problem with using shared memory is that we both require that the device support shared memory allocation and that the backend is able to detect shared memory dependencies coming from global variables. Using `urUSMContextMemcpyExp` does make this a blocking operation however, so I don't know what is better.